### DISCUSSION TITLE
Support S3 as a cache backend

### DISCUSSION THREAD

**Discussion by lyoung-confluent** [aquasecurity/trivy#6731] [Ideas]
### Description

Currently, `trivy` supports [an experimental redis cache backend](https://aquasecurity.github.io/trivy/v0.49/docs/configuration/cache/) instead of the local filesystem cache. It would be great to be able to also have support for S3 as a cache backend as it is ubiquitous and cheap. Interestingly, there is already a partial S3 cache implementation found inside the `trivy` repository: [pkg/fanal/cache/s3.go](https://github.com/aquasecurity/trivy/blob/main/pkg/fanal/cache/s3.go).

### Target

None

### Scanner

None

**Comment by lyoung-confluent** [comment]
PR: https://github.com/aquasecurity/trivy/pull/6732

**Reply by knqyf263** [reply]
Duplicate of https://github.com/aquasecurity/trivy/pull/3819

**Reply by lyoung-confluent** [reply]
Interesting, thanks for linking that, I did find #3820 (and meant to link it here) but I missed the mention of that PR in the timeline where some discussion had occurred previously. Is [this comment](https://github.com/aquasecurity/trivy/pull/3819#issuecomment-1477379922) still relevant regarding whether this feature would be accepted into mainline?

> Hi @zetaab , please excuse my delayed response. I'm sorry you spent time working on this before validating the issue but we are not planning to maintain more cache backends in Trivy. The s3 backend code that you found is a leftover of legacy implementation, and might get removed in a future refactoring.

For what it's worth I have added a relatively extensive set of tests to my PR and refactored the code, it's not quite the same as #3819 which seemed to just focus on making the existing implementation accessible.

**Comment by lyoung-confluent** [comment]
Self-closing per: https://github.com/aquasecurity/trivy/pull/6732#issuecomment-2209907787
